[1단계 - 출석 구현] 비타(오연택) 미션 제출합니다.

taek2222 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
 - [프롤로그](https://prolog.techcourse.co.kr/studylogs/3922)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

프리코스 때부터 습관 처럼 여겨져온 **객체지향 생활체조**를 실천했었고, 이번 미션에서도 최대한 지키기 위해 페어와 함께 노력했습니다. 🔥
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

시간내어 리뷰해 주셔서 진심으로 감사합니다. 🙇🏻

제가 부족한 점이 있다면 깊이 있게 고민하고 개선하고 싶습니다. 사소한 부분이라도 피드백이 있다면 짚어주시면 감사하겠습니다. 🔥

### Q. TDD 커밋 단위
TDD 연습을 시작하면서, Red, Green, Refactor의 단계를 기준으로 반복하는 과정에서 커밋 단위를 어떻게 쪼개야 할지 고민이 되었습니다. 각 단계에서 커밋을 분리하는 것이 바람직한지, 아니면 여러 단계를 한 번에 묶어서 커밋하는 것이 더 효과적인지 궁금합니다. 😢 

추가로, 커밋할 때 Red, Green, Refactor 을 명시하는지 궁금합니다. 🤔

---

### Q. TDD의 한계
TDD 연습을 하다 보면 기능이 어느 정도 완성된 후 점진적으로 합치는 과정에서 몇 가지 어려움을 겪었습니다.

- 필요한 메서드를 추가로 정의하게 되었을 때, 바로 테스트 코드를 먼저 작성해야 하는지 궁금합니다.

- TDD 방식으로 설계하다 보면 예상치 못한 변수나 파라미터 변경이 발생하는 경우가 있습니다. 이러한 변경이 기존 테스트에도 영향을 미쳐 수정해야 하는 경우가 많은데, 이를 최소화할 방법이 있을까요?

- 설계를 더 꼼꼼하게 하면 이러한 문제를 줄일 수 있을까요? 혹은 다른 접근법이 있을까요?

TDD를 진행하면서 구현 시간이 예상보다 증가하는 경우가 있어, 효율적인 접근법이 무엇인지 고민이 됩니다. 이에 대한 피드백을 주시면 감사하겠습니다!

### 🎯 집중 요청

- 항상 네이밍에 최대한 신경쓰려고 합니다! 네이밍 선택에 개선점이 있으면 무엇이든 짚어주시면 감사합니다! ☺️

---

추가적인 질문은 코드에 코멘트로 남겨두겠습니다! 🙇🏻

taek2222 committed:
docs(readme): 기능 구현 목록 초안 작성

taek2222 committed:
docs(readme): 기능 구현 목록 초안 수정

- 마크 다운 형식의 오해로 취소선 및 구분선 삭제했습니다.

taek2222 committed:
test(AttendanceWarningTypeTest): 테스트 추가

- 출결 상황에 따라 알맞은 경고를 반환하는 테스트를 추가 했다.

taek2222 committed:
feat(AttendanceWarningType): 타입 조회 기능 추가

- 출결 상황에 알맞게 타입을 찾아 반환한다.

- 단, 3회 지각은 1회 결석으로 처리한다.

taek2222 committed:
test(AttendanceWarningTypeTest): 테스트 리팩토링

- 출결 여러 상황에 알맞는 파라미터 테스트 형태로 변경했습니다.

taek2222 committed:
docs(readme): 구현 목록 체크

KJungW committed:
test(AttendanceStatusTypeTest): 테스트 추가

- 임계값에 해당하는 출결 상태를 반환한다

KJungW committed:
feat(AttendanceStatusType): 기능 추가

- AttendanceStatusType 생성 및 조회 구현

KJungW committed:
docs(readme): 구현 완료 기능 체크

taek2222 committed:
test(EducationTimeTest): 테스트 추가

- 월요일에 알맞은 지각 시간을 반환하는 테스트를 설계했다.

taek2222 committed:
feat(EducationTime): 지각 시간 계산 기능 추가

- 등교 시각과 시작 시각을 비교해 지각 시간을 반환한다.

taek2222 committed:
test(EducationTimeTest): 테스트 추가

- 요일 별로 여러 시간대의 파라미터 테스트를 추가했다.

taek2222 committed:
fix(EducationTime): 이전 시간 출석 버그 수정

- 절댓값으로 잘못된 차이를 계산해 출석한 경우의 차이값을 0으로 반환하게 했다.

KJungW committed:
test(CampusTimeTest): 테스트 추가

- 캠퍼스 운영시간이 아니라면 예외를 발생한다

KJungW committed:
feat(CampusTime): 캠퍼스 시간 구현

- 캠퍼스 운영 시간인지 검증하는 기능을 구현한다.

KJungW committed:
docs(readme): 구현 기능 목록 수정

- 구현된 기능 체크 표시
- 예외 상황 추가

KJungW committed:
test(CampusTimeTest): 테스트 리펙토링

- "캠퍼스 운영시간이 아니라면 예외를 발생한다."
  테스트 케이스 추가

taek2222 committed:
test(HolidayTest): 테스트 추가

- 등교일이 아닌 경우를 확인할 수 있는 테스트를 설계했다.

taek2222 committed:
feat(Holiday): 공휴일 검증 기능 추가

taek2222 committed:
test(HolidayTest): 테스트 추가

- 등교일 아닌 경우의 파라미터 테스트를 추가했다.

- 휴일 추가가 중복될 경우 예외 발생을 추가했다.

taek2222 committed:
feat(Holiday): 공휴일 추가하는 기능 추가

- 확장성을 고려해 공휴일을 추가하는 메서드를 추가했다.

- 공휴일도 휴일로 판정될 수 있도록 리팩토링 했다.

KJungW committed:
test(AttendanceTest): 테스트 추가

- 기본 날짜와 시간으로 출석을 생성할 수 있다

KJungW committed:
feat(Attendance): 디폴트 출석 기록 생성

- 입력된 날짜에 해당하는 디폴트 출석 기록을 생성한다.

KJungW committed:
test(AttendanceTest): 테스트 추가

- 기본 날짜와 시간으로 출석을 생성할 수 있다

KJungW committed:
feat(Attendance): 출석 시간 업데이트 구현

- 입력된 시간으로 출석 시간을 업데이트한다

KJungW committed:
test(AttendanceTest): 테스트 추가

- 출석 시간을 수정할 수 있다

taek2222 committed:
test(AttendancesTest): 테스트 추가

- 크루가 추가될때 기본 출석이 생성되는 테스트를 구현했다.

taek2222 committed:
feat(Attendances): 크루를 추가하는 기능

- 크루를 기본적으로 생성하는 로직을 설계했다.

- 크루 이름으로 찾을 수 있는 일정을 반환하는 메서드도 설계했다.

KJungW committed:
refactor(Attendance): 날짜 생성기 적용

- 테스트하기 어려운 날짜 생성 부분 분리

KJungW committed:
test(AttendancesTest): 테스트 수정

- 모킹 날짜 생성기를 주입해서 테스트하도록 수정한다.

taek2222 committed:
test(AttendanceTest): 테스트 추가

- 동일 날짜 여부에 따라 반환하는 값을 추가한다.

taek2222 committed:
feat(Attendance): 날짜 동일 여부 메서드 추가

taek2222 committed:
feat(Attendances): 출석 등록 기능 구현

- 애플리케이션 셋팅을 위해 설정을 주입했습니다.

- 입출력 부분을 정의해 비지니스 로직에 추가했습니다.

KJungW committed:
refactor(AttendanceManager): 기존 클래스명 수정

- Attendances를 AttendanceManager로 수정

KJungW committed:
test(AttendancesTest): 테스트 추가

- 출석을 추가할 수 있다
- 해당 날짜의 출석을 찾을 수 있다

KJungW committed:
test(HolidayTest): 테스트 실패 수정

- 공휴일 세팅값을 추가해서 테스트가 성공하도록 수정한다

KJungW committed:
refactor(Attendances): Attendances 적용

- 코드의 단순화를 위해 출석 리스트를 묶는 Attendances를 적용한다

KJungW committed:
feat(AttendanceService): 출석 수정 구현

- 기존의 출석 기록을 수정할 수 있다

taek2222 committed:
feat(attendances.csv): 출석 초기 데이터 생성

taek2222 committed:
feat(FileUtil): 파일 입력 유틸 추가

taek2222 committed:
feat(AttendanceService): 초기 크루 출석 설정 및 적용

KJungW committed:
feat(AttendanceService): 회원의 출석기록 출력

- 회원의 출석기록을 출력한다

taek2222 committed:
feat(AttendanceWarningType): 출석 이름 타입 추가

taek2222 committed:
feat(AttendanceService): 크루 출석 목록에 데이터 추가

KJungW committed:
feat(AttendanceService): 제적 위험자 조회 구현

- 제적 위험성이 있는 학생들을 조회한다

taek2222 committed:
docs(*): TodoList 주석 작성

taek2222 committed:
refactor(Attendances): 출석 저장 형태 Map -> 필드 전환

- 기능 변동으로 4번 기능의 오류로 주석처리 했다.

taek2222 committed:
refactor(Attendances): 오늘 날짜 까지만 기록 조회로 변경

- 출석 기록을 오늘 날짜까지만 출력할 수 있도록 변경했다.

taek2222 committed:
docs(readme): 기능 구현 완료 체크

KJungW committed:
feat(AttendanceManager): 제적 위험자 정렬

- 제적 위험자을 요구사항에 주어진 기준에 따라 정렬한다

KJungW committed:
docs(readme): 완료된 요구사항 체크

taek2222 committed:
refactor(AttendanceController): 재실행 기능 추가

- 필드 관리를 위해 메뉴 커맨드를 Enum으로 분리했다.

- 일치하지 않은 커맨드는 에러가 발생하도록 기능을 추가했다.

KJungW committed:
feat(AttendanceController): 메뉴 반복 입력 추가 적용

- 메뉴의 모든 기능에 대해 메뉴 반복 입력 추가 적용

KJungW committed:
feat(AttendanceManager): 등록되지 않은 닉네임 처리

- 등록되지 않은 닉네임이 임력될 경우 예외 처리

KJungW committed:
refactor(AttendanceManager): 등록된 크루 확인 수정

- 등록된 크루가 아닌 경우 true를 리턴한다.

KJungW committed:
feat(Attendance): 이미 출석된 경우 처리

- 이미 출석되었는데 출석을 시도하는 경우 예외처리

KJungW committed:
fix(AttendanceController): 메뉴 반복 버그 수정

- 무한 반복 버그 수정

taek2222 committed:
feat(InputView): 최소한 검증 로직 추가

- 검증 로직 추가로, 반환 값을 변환 형태로 반환하도록 변경했다.

taek2222 committed:
feat(AttendanceInitService): 초기 설정 서비스 분리

- DateTime 파싱 유틸 클래스를 정의해 에러까지 처리하도록 했다.

KJungW committed:
refactor(AttendanceService): 날짜와 시간 파싱에 에외처리 추가

- 날짜와 시간 파싱에 예외 처리를 추가한다.

taek2222 committed:
refactor(AttendanceService): 오늘까지 기록 반환 로직 도메인으로 변경

KJungW committed:
refactor(AttendanceService): 메서드 분리

- 코드의 단순화를 위해 메서드 분리를 수행한다

KJungW committed:
docs(readme): 구현 완료한 기능 체크

KJungW committed:
refactor(AttendanceManager): 코드 단축

- 스트림을 활용해 기존의 코드를 단축한다

taek2222 committed:
fix(AttendanceManager): 마지막 날까지 생성 안되는 버그 수정

taek2222 committed:
refactor(Attendances): 출석 줄바꿈 개선

taek2222 committed:
fix(AttendanceService): 출력 편집 수정

taek2222 committed:
refactor(view): 출력 형식에 알맞게 변경

taek2222 committed:
refactor(dto): dto 패키지 분리 및 네이밍 간소화

- response 해당되는 패키지를 분리해 네이밍을 최소화 했습니다.

taek2222 committed:
refactor(InputValidator): final 및 객체 생성 방지

- Util 성격에 가까워 private 생성자로 방지했습니다.

taek2222 committed:
style(DateGenerator): 공백 제거

taek2222 committed:
refactor(FileUtil): 생성 방지 및 상수 분리

taek2222 commented on src/main/java/attendance/controller/AttendanceMenu.java:
```diff
@@ -0,0 +1,24 @@
+        this.command = command;
+    }
+
+    public static AttendanceMenu find(String input) {
+        return Arrays.stream(values())
+                .filter(menu -> menu.command.equals(input.toUpperCase()))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("[ERROR] 메뉴의 알맞은 커맨드를 입력해주세요."));
+    }
```
Q. `find` 의미로 기능을 충분히 이해할 수 있을까요? 🤔

이번 페어 프로그래밍의 '랜디' 의견으로 `find` 명칭만 네이밍을 진행했습니다. 너무 함축적이고, 생략된 느낌이 있어 보여서 충분히 이해할 수 있을지 의문입니다. (아직 어색해서 그런지.. 🤣)

jinyoungchoi95 replied:
함축의 문제가 아니라 `find`는 "찾다"입니다. 저장되어있는 값을 찾는게 아니라 이건 파싱하는 로직이 아닌가요?

taek2222 replied:
다시 생각해보니 "파싱"에 의미가 가깝군요.. 🤔 

그렇다면, 메서드 네이밍을 `parse()`로 할지, `parseMenuByCommand()`로 할지 고민 중입니다. 

parse()처럼 간결한 네이밍이 이 기능을 충분히 설명할 수 있을까요?
아니면 parseMenuByCommand()처럼 좀 더 구체적인 네이밍이 적절할까요?

네이밍을 결정할 때 고려해야 할 요소가 있다면 조언 부탁드립니다! 🙇🏻

jinyoungchoi95 replied:
> 줄여 쓰지 않는다(축약 금지).

과해도좋으니 축약하지말고 최대한 의미를 잘 전달할 수 있도록 지으면 좋습니다. 처음보는사람이 봐도 메소드명만 봐도 와닿을 수 있으면 된다고 생각해요.

taek2222 committed:
refactor(DateTimeParser): 메시지 상수 분리 및 Final 키워드 삽입

taek2222 commented on src/test/java/attendance/domain/AttendanceTest.java:
```diff
@@ -0,0 +1,73 @@
+
+class AttendanceTest {
+
+    @DisplayName("기본 날짜와 시간으로 출석을 생성할 수 있다.")
+    @Test
+    void 기본_날짜와_시간으로_출석을_생성할_수_있다() {
+        assertThatCode(() -> new Attendance(LocalDate.MIN))
+                .doesNotThrowAnyException();
+    }
```
Q. 에러가 발생하지 않는 생성자에서도 올바르게 작동하는 지 테스트를 해도 될까요?

객체가 정상적으로 생성되었는지 확인하기 위해, `getter` 사용으로 값을 확인할 수 있지만 최대한 사용을 하지 않기 위해서 설정하지 않았습니다. 그래서 생각한 방법이 위와 같이  _"에러가 발생하지 않는가?"_ 로 판단하는데 막상 구현하다보니 _"이것도 테스트를 해야할까?"_ 라는 생각을 했습니다.

jinyoungchoi95 replied:
객체가 잘 생성되는 것으로 충분히 테스트의 목적은 달성하고 있는 것 아닐까요? TDD의 과정을 생각해보면 잘 생성되는 케이스도 충분히 고려해볼만하지 않을까요?

taek2222 replied:
의견 주셔서 감사합니다! ☺️ 추가적으로 고민 후 보완해보겠습니다!

jinyoungchoi95 commented on README.md:
> Q. TDD 커밋 단위
> TDD 연습을 시작하면서, Red, Green, Refactor의 단계를 기준으로 반복하는 과정에서 커밋 단위를 어떻게 쪼개야 할지 고민이 되었> 습니다. 각 단계에서 커밋을 분리하는 것이 바람직한지, 아니면 여러 단계를 한 번에 묶어서 커밋하는 것이 더 효과적인지 궁금합니다. 😢

각 단계를 분리하는 것과 한번에 묶는 것의 각각의 장단점은 무엇인가요? 구체적으로 먼저 왜 그것들이 좋은지를 정리해주시고나서 같이 이야기해보면 좋을 것 같아요.

taek2222 replied:
## 각 단계 커밋
### 장점 
- TDD 과정의 세부적인 흐름을 추적할 수 있다.
- 필요한 경우 특정 지점으로 `롤백`이 용이하다.

### 단점
- 너무 자주 커밋하면 시간이 다소 소요될 수 있다.
- 커밋이 많아지면 히스토리가 복잡해지고 한 번에 이해하기 어려울 수 있다.

## 한 묶음 커밋
### 장점
- 기능의 구현 과정을 한 번에 확인할 수 있다.
- 커밋 횟수를 줄여 TDD 진행 시간을 단축할 수 있다.
- 보다 깔끔한 커밋 히스토리를 유지할 수 있다.

### 단점
- 중간 과정이 생략되면 특정 `이슈` 발생 시 원인 파악이 어려울 수 있다.
- `롤백` 시 원하는 지점을 찾기가 어려울 수 있다.

위와 같이 생각해 보았습니다. 정리하면서도 여러 고민이 있었지만, 시간의 여유가 있을 때?는 `각 단계 커밋`이 적절할 거 같습니다. 다만, 시간이 부족할 경우 `한 묶음 커밋`이 좋은 방향일 거 같다는 생각을 했습니다.

jinyoungchoi95 replied:
롤백의 지점을 정확하게 컨트롤할 수 있다는 것만으로도 단계별로 커밋하는 이유로 충분합니다. 돌아갈 수 있는 포인트는 상세하면 상세할수록 좋고, 조금이라도 커지면 원하는 지점으로 못돌아가니까요. 시간적 여유가 있다 없다는 사실 아직 익숙하지 않기 때문이라 생각해요. 

TDD나 클린코드, 단위테스트 커버리지 등 시간적인 이유를 단점으로 많이 삼는데 많이 연습하고 익숙한 사람의 입장에서는 시간적인 이유가 단점이 되는 것 같지는 않습니다. 지금 시간이 오래걸린다고 단점으로 삼지 말고 최대한 익숙해지려고 연습해보세요. 그러려고 있는 것이 1레벨의 많은 미션들이니까요.

taek2222 replied:
가장 좋은 코멘트를,, 리뷰 제출 직전 마지막에 봤습니다! 😢

따뜻한 말씀 감사합니다! 열심히 해보겠습니다 🔥

jinyoungchoi95 commented on README.md:
```diff
@@ -1,3 +1,273 @@
 # java-attendance
 
-출석 미션 저장소
+## :sparkling_heart: 프로젝트 소개
```
> 필요한 메서드를 추가로 정의하게 되었을 때, 바로 테스트 코드를 먼저 작성해야 하는지 궁금합니다.

네. 테스트코드를 작성하는 것에 의의를 두지말고 실패 케이스를 만든다에 초점을 맞추어보세요.

왜 실패 케이스를 먼저 만드는 것이 프로그램을 만드는데 도움이 될까요? 실패 케이스를 왜 먼저 만드느냐를 고민하는 것이 우선이기 때문에 필요한 메서드의 정의는 여기서 중요하지 않아보여요. (메서드의 정의에 실패 케이스가 포함되기는 하지만 메서드의 정의라는 이야기는 너무 포괄적인 것 같아요.)

jinyoungchoi95 commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+웹 백엔드 7기 레벨1 출석 미션을 구현한 프로젝트입니다.
+
+## :dart: 구현 기능 목록
```
> TDD 방식으로 설계하다 보면 예상치 못한 변수나 파라미터 변경이 발생하는 경우가 있습니다. 이러한 변경이 기존 테스트에도 영향을 미쳐 수정해야 하는 경우가 많은데, 이를 최소화할 방법이 있을까요?

테스트는 변경의 여파가 있을 때 실패하라고 만든 것입니다. 코드가 변경이 되어 영향이 가는 것이 당연한데 기존 테스트가 잘 돌아간다면 그게 오히려 더 큰 문제가 아닐까요?

taek2222 replied:
굉장히 제가 생각하고 있었던 잘못된 테스트 개념을 깨닫게 해주셨습니다..! 🫢 _"테스트는 변경의 여파가 있을 때 실패하라고 만든 것입니다."_ 와닿았습니다. 감사합니다!☺️

jinyoungchoi95 commented on README.md:
```diff
@@ -1,3 +1,273 @@
+
+## :dart: 구현 기능 목록
+
+## 출석 파일 입력
```
> TDD를 진행하면서 구현 시간이 예상보다 증가하는 경우가 있어, 효율적인 접근법이 무엇인지 고민이 됩니다. 이에 대한 피드백을 주시면 감사하겠습니다!

TDD가 익숙하지 않고 처음 하는 것이라 그렇습니다. 해보지 않았던 방식이 오래걸리는 것은 당연하고 앞으로의 미션은 TDD를 익숙하게하는 연습이 되는 것이구요.

효율적인 접근법은 사실 사람마다 설계를 하는 방법이 다르기 때문에 어떤 것을 하라라는 이야기는 못할 것 같고, 말씀드린대로 앞으로의 미션에서 TDD를 하면서 비타가 생각하는 효율적으로 진행해나가는 방법을 찾아가는 것이 좋을 것 같아요.

taek2222 replied:
답변 감사합니다. 

주변 크루와 지인들로부터 _"TDD 생각보다 쉽지 않네..", "구현 시간이 오래 걸리는 거 같네.."_ 등의 의견이 들리니 더욱 걱정했던 거 같습니다. 😢

앞으로 지속적으로 연습해서 TDD와 친해지고 🔥, 효율적인 방법을 탐색해 보겠습니다! 🤔

jinyoungchoi95 commented on src/main/java/attendance/service/AttendanceInitService.java:
1. 여기는 비즈니스 로직이 아닌가요? 테스트가 필요한 것 아닐까요?
2. 우테코의 레벨 1 미션은 "자바"에 대한 학습입니다. 자바를 공부할때 layered architecture가 필요한 것이 맞을까요? Service, Controller 보다는 정말 잘 사용할 수 있고 기능을 잘 설명할 수 있는 객체로 만들어보세요. 레벨 1때에 얻어가야할 것들이 산더미이기 때문에 레벨 2에서 배울 수 있는 내용들은 잊고 순수하게 자바만 사용해보는 개발자라고 생각하고 설계해보세요. (코딩 테스트에 controller를 안쓰는 것처럼요.)

콘솔 게임이라는 설계기준을 보면 Service, Controller를 바라보면 무슨 일을 하는 객체인지 잘 안와닿네요.

taek2222 replied:
제가,, 레벨 1 미션의 의도를 인지하지 못한 점이 큰 실수였던 거 같습니다. 😢

말씀해주신 내용처럼 `자바`에 대한 학습을 우선 시 해야하는데, 시작부터 아키텍쳐 고민하고 그런 시간이 레벨 1 미션에 적용하기에는 시간이 아깝다는 생각을 했습니다.

한 번 객체로 최대한 변환 적용해 기능을 잘 설명해보겠습니다! ☺️

jinyoungchoi95 commented on src/main/java/attendance/dto/response/WarnedStudent.java:
dto의 정의를 생각해본다면 dto에 `calculateScore`와 같은 도메인 로직이 들어가는 것이 맞을까요? 그리고 불변한 객체를 잘 설계하고 그대로 view로 넘겨도 될 것 같은데 dto를 쓸 필요가 있나요? 위에서 말씀드린 것처럼 layered architecture라는 것을 빼고 고민해보시면 사실 dto는 현재 미션에서 필요없는 존재인 것 같아요.

taek2222 replied:
지적해주신 내용 정말 감사합니다. 이전까지의 생각은 _"`Getter`는 무조건 선언하지 말아야지"_ 라고 생각해 `getter` 사용을 하지 않기 위해 `DTO`를 선언했으나, 이번 코멘트 남겨주신 내용으로 생각의 흐름을 바꿀 수 있었습니다.

`getter` 사용을 지양하라는 내용은 _"절대 선언하지 마라!"_ 아닌, _"객체 관련 로직을 최대한 캡슐화해 적용하라"_ 의 시선으로 바꿀 수 있었습니다.

감사합니다! 항상 고민했던 부분이 깔끔하게 정리된 느낌입니다! 😲

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,43 @@
+
+    public boolean isAlreadyCheck() {
+        return !dateTime.toLocalTime().equals(LocalTime.MIN);
+    }
```
domain에도 테스트가 안된 코드들이 많이 보여요. TDD를 하면 테스트코드를 먼저 작성하게 될건데 테스트가 있었어야하지 않을까요?

taek2222 replied:
마지막날, 구현한 메서드를 종합하면서 생긴 긴급하게 만들어 테스트를 깜빡했습니다.. 😢

항상 추가된 기능은 TDD 단계의 Red를 먼저, 작성하는 습관을 길들이겠습니다! 🔥

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,43 @@
+public class Attendance {
+
+    private LocalDateTime dateTime;
+    private AttendanceStatusType status;
```
도메인이 불변이 되고 있지 않으면 믿고 `Attendance`를 사용할 수 있을까요? 불변하게 설계하는 방법을 고려해보면 좋을 것 같아요.

taek2222 replied:
1. 필드의 `final` 키워드 삽입
``` java 
    private final LocalDateTime dateTime;
    private final AttendanceStatusType status;
```

2. 생성자 방식 변경
``` java
    public Attendance(LocalDateTime dateTime) {
        this.dateTime = dateTime;
        
        int overTime = EducationTime.calculateOverTime(dateTime);
        this.status = AttendanceStatusType.find(overTime);
    }
```

2. Update 메소드 변경
``` java
    public void updateTime(LocalDateTime dateTime) {
        new Attendance(dateTime);
    }
    혹은, 외부에서 Update 메서드로 새로 생성
```

주신 피드백 내용을 바탕으로 개선 방향을 찾았습니다.. 불변 꼭 메모해두겠습니다. 📝

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,43 @@
+
+    public AttendanceRecord createResponse() {
+        return new AttendanceRecord(dateTime, status.getName());
+    }
```
(dto가 필요없다라고 이야기했지만 코드에서 dto가 있는 현재 상태에 대해서 말씀드리면) domain이 dto에 의존하는 것이 맞나요? dto의 요구사항(화면에서 필요한 데이터)가 변경되면 도메인에 영향이 가는 것 아닐까요?

taek2222 replied:
단지, getter 사용을 지양하고자 했던 방식이었습니다. 지금 돌이켜보니 dto 사용으로 각 model과 view의 영향을 최소화 했다고 생각했지만.. 큰 착각이었던 거 같습니다. 🤔

jinyoungchoi95 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,58 @@
+        return new AttendanceRecordUntilToday(responses);
+    }
+
+    private int calculateStatusUntilYesterday(AttendanceStatusType status, LocalDate date) {
```
메소드를 직역하면 "어제까지의 상태를 계산한다"인데 어떤 기준으로 어떤걸 계산하는지 너무 추상화되어있는 것 아닐까요?

taek2222 replied:
코멘트를 보고서 네이밍을 고민했을 때,

1. `calculateAttendanceStatusUpToYesterday`
2. `getAttendanceStatusUntilYesterday`

이렇게 두가지 떠오릅니다.. 만약, 위 메서드 이름이 적절하다고 하면 궁금한 점이 있습니다.

항상 네이밍을 할 때, 길이를 어느정도로 제한을 둬야할 지 모르겠습니다..! 🤔 지금 고민한 이름도 상당히 길어보여서 흠.. 😢

jinyoungchoi95 replied:
제한을 두지말고 네이밍해보세요. 우테코 기간은 연습하는 기간이니 축약하는 것보다 자세한 것이 좋습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/Holiday.java:
```diff
@@ -0,0 +1,38 @@
+
+    private final List<LocalDate> publicHolidays = new ArrayList<>(); // todo : 공휴일 네이밍 의문, 객체 생성 과정 의문
+
+    public void addHoliday(LocalDate holiday) { // todo : 파라미터 네이밍 고민, 크리스마스 추가
```
사용되는 곳이 없어보이네요.

taek2222 replied:
2024, 12월 기준으로 프로그램에서 크리스마스 공휴일이 추가되었어야 했는데, 만들고 적용하지 못했던 거 같습니다! 👀

jinyoungchoi95 commented on src/main/java/attendance/domain/Holiday.java:
```diff
@@ -0,0 +1,38 @@
+
+public class Holiday {
+
+    private static final DayOfWeek SATURDAY = DayOfWeek.SATURDAY;
+    private static final DayOfWeek SUNDAY = DayOfWeek.SUNDAY;
```
enum으로 이미 상태가 정의되어있는데 이걸 상수로 뺄 필요가 있을까요? 차라리 공휴일이라는 토요일과 일요일을 모두 들고 있는 상수가 나아보여요.

taek2222 replied:
헉,,! 이미 존재한 상수를 또.. 정의했군요. 주의하겠습니다! 🫡

jinyoungchoi95 commented on src/main/java/attendance/utility/DateTimeParser.java:
```diff
@@ -0,0 +1,45 @@
+
+public final class DateTimeParser {
+
+    private static final String PATTERN = "yyyy-MM-dd HH:mm";
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern(PATTERN);
```
```suggestion
    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
(반영하지 않으셔도 됩니다.) 상수의 목적은 모든 매직넘버/리터럴을 없앤다기보다는 의미를 잘 전달하기 위함이에요. 그래서 사실 모든 것을 상수화한다기보다는 목적에 따라서 상수화해주시면 좋습니다.

taek2222 replied:
의견 감사합니다! 상수의 목적을 인지하고 적용해보겠습니다! 🚀

jinyoungchoi95 commented on src/main/java/attendance/utility/FileUtil.java:
```diff
@@ -0,0 +1,36 @@
+            FileReader fileReader = new FileReader(file);
+            return new BufferedReader(fileReader);
+        } catch (FileNotFoundException e) {
+            throw new IllegalArgumentException();
```
실패하는 원인에 대한 적절한 메시지를 남겨주어야하지 않을까요?

taek2222 replied:
헉,,! 맞습니다. 적용하겠습니다! 😲

jinyoungchoi95 commented on src/main/java/attendance/domain/EducationTime.java:
```diff
@@ -0,0 +1,35 @@
+
+public enum EducationTime {
+    MONDAY(LocalTime.of(13, 0, 0), LocalTime.of(18, 0, 0)),
+    NOT_MONDAY(LocalTime.of(10, 0, 0), LocalTime.of(18, 0, 0)); // todo : 추후 네이밍 고민
```
1. todo나 사용하지 않는 메소드들은 모두 제거해주세요.
2. 네이밍 고민이 필요하다하셔서 남겨보면, 월요일이라는 것을 명시하는 것이 맞나요? 1시 출근이 화요일이 되면 여기도 수정되어야하겠네요.

taek2222 replied:
1. 앞으로는 주의해서 리뷰 요청하겠습니다! 🫡

2. 의견 감사합니다..! 페어랑 함께 고민하면서 좋은 방향이 떠오르지 않아 위와 같이 결정하고 넘어간 거 같습니다! 😢

jinyoungchoi95 reviewed (changes requested):
안녕하세요 비타 😄

출석 미션의 리뷰어를 맡은 오리입니다.

첫 미션 잘 구현해주셨네요. 다음 단계 넘어가기 전 같이 이야기하면 좋을 내용 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

taek2222 reviewed (commented):
소중한 리뷰 감사합니다! 충분한 학습을 진행하고, 리팩토링 해보겠습니다! 🫡

오늘도 행복한 하루 보내십쇼! ☺️

taek2222 committed:
refactor(AttendanceController): try ~ catch 삭제

- 요구사항의 [ERROR] 출력후 재입력 기능은 없어 삭제했습니다.

taek2222 committed:
refactor(AttendanceInit): InitService 삭제 및 Init 도메인 생성

- Service 레이어 보다는 객체 사이에 역할에 집중할 수 있도록 변경했습니다.

taek2222 committed:
refactor(AttendanceController): service 레이어 삭제

taek2222 committed:
test(AttendanceSystemTest): 테스트 추가

- 크루 출석 확인하는 기능의 테스트를 설계했다.

taek2222 committed:
feat(AttendanceSystem): 출석 체크 기능 구현

- 불변 유지를 위해 Attendance 부분을 final로 변환하고, 생성자 형태를 변경했다.

- Attendances 업데이트 기능을 새로 정의했다. 불변을 위해 기존 일정을 삭제하고 삽입하는 형태로 변경했다.

taek2222 committed:
test(*): Attendance 생성자 변동 리팩토링

taek2222 committed:
test(AttendanceSystemTest): 테스트 리팩토링

- 날짜를 mocking해 지정된 일자 형태로 변경했습니다.

- 비즈니스 구현 중 변동되는 부분을 수정했습니다.

taek2222 committed:
test(AttendanceSystemTest): 테스트 리팩토링

- 파라미터 테스트로 여러 시간의 출석을 확인했다.

taek2222 committed:
test(AttendanceSystemTest): 테스트 추가

- 존재하지 않은 닉네임으로 출석 체크시 에러 발생하는 검증을 추가했다.

taek2222 committed:
test(AttendancesTest): 테스트 추가

- 이미 출석한 크루의 출석 확인할 경우 에러가 발생하는 검증을 추가했다.

taek2222 committed:
feat(Attendances): 이미 출석한 경우 검증 추가

taek2222 committed:
feat(AttendanceSystem): 등록되지 않은 크루 검증 추가

taek2222 committed:
test(AttendanceSystemTest): 테스트 수정

- 등록 크루 검증 로직 따로 분리로 변경했다.

taek2222 committed:
fix(AttendanceSystem): 크루 검증 로직의 잘못된 부분 수정

taek2222 committed:
feat(AttendanceController): 날짜 검증 로직 추가

taek2222 committed:
feat(OutputView): 객체로 출력하는 로직 추가

taek2222 committed:
test(AttendanceSystemTest): 테스트 추가

- 출석 수정하는 테스트하는 검증을 추가했다.

taek2222 committed:
feat(AttendanceSystem): 출석 수정 기능 구현

taek2222 committed:
test(AttendanceSystemTest): 테스트 추가

- 크루의 출석 데이터가 없는 수정일 경우 에러가 발생하는 검증을 추가한다.

taek2222 committed:
refactor(Attendances): 수정할 데이터 없는 경우 에러 추가

taek2222 committed:
refactor(AttendanceManager): System 객체 삭제 및 처리 이동

taek2222 committed:
refactor(AttendanceManager): 출석 등록 및 수정 로직 리팩토링

taek2222 committed:
refactor(AttendancesTest): 테스트 리팩토링

- 기존 로직의 변경으로, 여러 메소드를 수정했다.

taek2222 committed:
refactor(AttendanceController): System 객체 삭제로 Manager 변경

taek2222 committed:
refactor(AttendanceController): 출석 수정 기능 변경 완료

taek2222 committed:
test(AttendanceManagerTest): 테스트 추가

- 크루의 전날까지 기록을 반환하는 검증을 테스트한다.

taek2222 committed:
feat(AttendanceManager): 전날까지의 기록 반환 추가

taek2222 committed:
fix(AttendanceManagerTest): 테스트 수정

- 실수로 적은 값을 수정했다.

taek2222 committed:
refactor(OutputView): 출력문 객체로 변경

taek2222 committed:
test(AttendanceTest): 테스트 추가

- 출석 시간을 비교해 반환하는 값을 검증하는 테스트를 추가했다.

taek2222 committed:
feat(Attendance): 비교하는 기능 추가

taek2222 committed:
test(AttendanceTest): 테스트 리팩토링

- 여러 상황을 확인하기 위해 출석 시간 비교 파라미터 테스트 추가했다.

taek2222 committed:
feat(Attendances): 출석 리스트 반환 시 정렬 추가

taek2222 committed:
test(AttendanceManagerTest): 테스트 추가

- 출결 상태를 확인 하는 검증 테스트를 추가했다.

taek2222 committed:
test(AttendanceStatusTest): 테스트 추가

- 출석 객체를 생성하는 로직을 검증한다.

taek2222 committed:
feat(AttendanceStatus): 출결 상황 보관 객체 생성

- 출석 결과를 바탕으로 상황을 보관하도록 생성했습니다.

taek2222 committed:
test(AttendanceStatusTest): 테스트 리팩토링

- 가독성 향상을 위해 메서드 분리를 적용했다.

taek2222 committed:
feat(AttendanceStatus): 생성자 리팩토링 및 정적 팩토리 메서드 추가

taek2222 committed:
feat(AttendanceManager): 크루 추가, 출결 상황 반환 로직 변경

- 크루 추가를 오늘 기준까지로 생성되도록 변경했습니다.

- 크루의 출결 현황을 반환하는 로직을 정적 팩토리 메서드를 활용하는 쪽으로 변경했습니다.

taek2222 committed:
refactor(Attendance): 이미 출석을 MAX로 변경

- 기본 출석을 MAX로 설정해 다음과 같이 변경했습니다.

taek2222 committed:
refactor(AttendanceStateType): 클래스 네이밍 변경

- AttendanceStatus 네이밍 중복으로, Enum을 더 하위 단어로 변경했습니다.

taek2222 committed:
feat(OutputView): 출결 상황 메서드 정의

- 기본 출력을 MIN -> MAX로 변경했습니다.

taek2222 committed:
refactor(Attendances): 파라미터 삭제

- 오늘까지 생성되는 형태로 변경함에 따라 마지막 날만 제거하는 형태로 변경했습니다.

taek2222 committed:
refactor(AttendanceStatus): 출석 상태 생성 타입 변경

- 전날까지의 출석 현황을 리스트로 받기 위해 파라미터를 변경했습니다.

taek2222 committed:
test(AttendanceStatusTest): 테스트 리팩토링

- 반환 타입 변경으로 테스트 수정했습니다.

taek2222 committed:
test(AttendanceManagerTest): 테스트 추가

- 제적 위험자의 크루를 조회하는 검증을 추가했습니다.

taek2222 committed:
feat(AttendanceManager): 위험자 크루 반환 로직 추가

taek2222 committed:
test(AttendanceManagerTest): 테스트 추가

- 제적 위험자가 아닌 크루는 제외하고 반환하는 검증을 추가했습니다.

taek2222 committed:
feat(AttendanceManager): 제적 위험자 필터링 추가

- 제적 위험자 대상이 아닌 크루는 제외하도록 필터링을 추가했습니다.

taek2222 committed:
fix(Attendances): 삭제 후 정렬 로직 변경

- 정렬 후 제외해야 하지만 처음에 삭제해 이상한 결과값이 도출되어 수정했습니다.

taek2222 committed:
refactor(*): 사용하지 않은 메서드, 클래스 삭제

- 도메인 중심 리팩토링으로 사용하지 않는 파일을 정리했습니다.

taek2222 committed:
refactor(Attendance): 메서드 네이밍, final 키워드, 순서 정렬

- 가독성 향상을 위해 메서드 네이밍을 변경했습니다.

- 불변의 명시를 위해 final 키워드를 삽입했습니다.

- private, public 메서드 순서를 정려했습니다.

taek2222 committed:
test(AttendanceTest): 부족한 테스트 보완

- 이미 출석, 출석 상태 동일 테스트를 새로 추가했습니다.

- 다른 부분에 부족한 파라미터 테스트를 추가해 설계했습니다.

taek2222 committed:
feat(AttendanceController): 위험자 조회 출력 연결

taek2222 committed:
test(AttendanceManagerTest): 테스트 추가

- 제적 위험자 출력 순서에 알맞게 순서 정렬 검증을 추가했습니다.

taek2222 committed:
feat(AttendanceManager): 크루 위험자 정렬 로직 추가

taek2222 committed:
refactor(AttendanceSystem): Controller 삭제

- 오직 자바 프로그램을 위해, 아키텍쳐의 의미 컨트롤러 레이어 삭제했습니다.

taek2222 committed:
test(AttendanceMenuTest): 테스트 추가

- 입력에 알맞는 메뉴를 반환하는 검증을 추가했습니다.

taek2222 committed:
refactor(AttendanceMenu): 명칭 일관성 및 상수 분리

taek2222 committed:
refactor(AttendanceSystem): 네이밍 및 메서드 분리 리팩토링

- 중복되는 의미의 네이밍을 간소화했습니다.

- 가독성 향상을 위해 메서드 분리했습니다.

- 명칭의 일관성을 위해 몇가지 변수와 메서드 네이밍을 수정했습니다.

taek2222 committed:
refactor(AttendanceInit): 기본 크루의 정보를 생성하는 로직 이동

taek2222 committed:
test(AttendanceManagerTest): 테스트 리팩토링

- 여러 생성자 및 파라미터 변동으로 바뀐 부분을 수정했습니다.

taek2222 committed:
refactor(OutputView): 주석 삭제 및 네이밍 수정

- 복수형의 잘못된 표기로 수정했습니다.

taek2222 committed:
refactor(FileUtil): 상수 분리 및 에러 메시지 정의

taek2222 committed:
feat(AttendanceInit): 공휴일(크리스마스) 추가

taek2222 committed:
test(HolidayTest): 테스트 추가

- 휴일인지 아닌지의 검사를 검증하는 테스트를 추가했습니다.

taek2222 committed:
refactor(Holiday): 중복 상수 제거 및 네이밍 수정

- 상수를 중복으로 정의하는 부분을 삭제했습니다.

- 네이밍을 더 직관적으로 바꾸기 위해 변경했습니다.

- 중복되는 로직을 메서드를 활용해 적용했습니다.

taek2222 committed:
refactor(EducationTime): 네이밍과 로직 최적화

- 의미가 불문명한 네이밍을 변경했습니다.

- overTime 계산 로직을 더욱 간단하게 변경했습니다.

taek2222 committed:
refactor(CampusTime): 파라미터 네이밍 및 로직 수정

taek2222 committed:
refactor(AttendanceWarningType): 주석 삭제 및 로직 간소화

taek2222 committed:
test(AttendanceStateTypeTest): 테스트 리팩토링

- 로직 간소화와 의미 없는 사용 괄호 삭제했습니다.

taek2222 committed:
refactor(AttendanceStateType): 로직 간소화

taek2222 committed:
feat(Attendance): 인터페이스 Comparable 정의

taek2222 committed:
test(AttendanceTest): 테스트 추가

- 날짜를 비교해 반환하는 메서드를 추가했습니다.

taek2222 committed:
feat(Attendance): 날짜 비교 메서드 추가

taek2222 committed:
refactor(Attendances): 특정 일자까지 반환 형태 변경

taek2222 committed:
test(AttendancesTest): 테스트 추가

- 출석을 삭제하는 테스트를 추가로 검증하도록 추가했습니다.

- 특정 일자까지의 출석 기록을 반환하는 메서드 검증도 추가했습니다.

taek2222 committed:
refactor(AttendanceStatus): 메서드 분리 및 로직 최적화

taek2222 committed:
refactor(AttendanceManager): 메서드 네이밍 변경

taek2222 committed:
refactor(*): Enum 타입의 네이밍 수정

taek2222 committed:
test(AttendanceStatusTest): 테스트 추가

- 비교 결과를 반환하는 값을 테스트 하는 것을 추가했습니다.

taek2222 committed:
fix(AttendanceStatus): 비교 값 잘못된 결과 해결

- 테스트 결과 잘못된 결과를 반환하는 것을 보고 수정했습니다.

taek2222 committed:
refactor(AttendanceInit): 접근제어자 수정

taek2222 committed:
feat(*): final 키워드 삽입

taek2222 commented on src/main/java/attendance/AttendanceSystem.java:
```diff
@@ -0,0 +1,126 @@
+        return AttendanceMenu.find(inputView.readMenuCommand());
+    }
+
+    private void processSystem(final AttendanceMenu menu, final LocalDate today) {
```
Q. 메서드를 정의할 때, 파라미터 정의에도 순서가 있나요? 🤔

리팩토링을 여러 해보면서 궁금한 점이 파라미터도 순서가 있는지 궁금해졌습니다. 
``` java
(final AttendanceMenu menu, final LocalDate today)
```

``` java
(final LocalDate today, final AttendanceMenu menu)
```

위처럼 파라미터 선언 순서 기준이 따로 있나요!?

jinyoungchoi95 replied:
사람마다 다릅니다. 먼저 사용되는 파라메터를 앞에 두기는합니다.

taek2222 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,43 @@
+
+public class Attendances {
+
+    private final List<Attendance> attendances = new ArrayList<>();
```
Q. 객체 생성 시점에 차이가 있나요? 🤔

위 예시처럼, 별도의 생성자가 필요 없는 경우, 필드에서 바로 객체를 생성해도 괜찮을까요?

jinyoungchoi95 replied:
자바에서 제공하는 기능이니 문제없다고 생각해요

taek2222 commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,75 @@
+import static attendance.domain.AttendanceStateType.LATE;
+import static attendance.domain.AttendanceStateType.values;
+
+public class AttendanceStatus implements Comparable<AttendanceStatus> {
```
Q. 객체 내부에 인터페이스를 구현해도 괜찮나요? 🤔

정렬 로직을 별도의 `Util` 클래스로 분리하는 방법과, 객체 내부에서 `Comparable`을 직접 구현하여 정렬 기준을 정의하는 방법 중 고민하다가 후자를 선택했습니다.

이 방식에 문제점은 없을까요? 더 나은 접근 방식이 있을까요? 🤔

jinyoungchoi95 replied:
1. util 클래스은 객체지향인가요?
2. 현재 방식이 왜 문제가 있다고 생각하시나요?

taek2222 replied:
> **util 클래스은 객체지향인가요?**

A. 유틸 클래스는 객체지향이 아닌, 절차지향의 성향을 띄고 있습니다. 질문 주신 내용을 토대로 자료를 조사한 결과, [유틸 클래스란?](https://velog.io/@ikjo39/%EC%9C%A0%ED%8B%B8-%ED%81%B4%EB%9E%98%EC%8A%A4%EB%9E%80) 로 통해 요구하는 행동을 객체로 변경하는 방향도 고민하게 되었습니다.

> **현재 방식이 왜 문제가 있다고 생각하시나요?**

A. 따로 문제가 될거라는 생각은 하지 않았습니다. 단순히, _**객체에 선언을 해도 괜찮을까?**_ 라는 의문만 들었던 거 같습니다.. 🤔

taek2222 commented on src/main/java/attendance/AttendanceSystem.java:
```diff
@@ -0,0 +1,126 @@
+        processRiskSearch(menu, today);
+    }
+
+    private void processCheck(final AttendanceMenu menu, final LocalDate today) {
```
Q. 축약을 주의하라 했지만, 위와 같은 경우는 괜찮을까요? 🤔

원래 메서드 이름은 `processAttendanceCheck`이었지만, 클래스 이름에 이미 `Attendance`가 포함되어 있어 중복이 신경 쓰여 변경을 고민하고 있습니다.

이 경우, 축약해도 무방할까요? 의견 부탁드립니다.

jinyoungchoi95 replied:
process라는 네이밍이 필요할까요? prefix로 다 달려있는데 단순히 `checkAttendance`면 되는 것 아닐까요?

taek2222 commented on src/test/java/attendance/domain/AttendanceManagerTest.java:
Q. 올바른 테스트 방식이 맞을까요? 🤔

검증이 필요한 메서드마다 테스트 코드를 작성하다 보니, 전체 코드량이 상당히 증가했습니다. 이렇게 작성하는 것이 일반적인 방식일까요, 아니면 더 효율적인 방법이 있을까요? 의견 부탁드립니다.

jinyoungchoi95 replied:
메소드명이 축약된다고 좋은 메소드명이 아니듯 코드량이 느는것이 안좋은 코드의 지표가 되지는 않습니다. 그리고 일반적으로 어떻게 테스트코드를 짠다는 정답도 없구요. 테스트 코드의 양보다는 객체의 역할이 무겁다고 느끼면 객체의 분리정도는 생각해볼 수 있겠네요.

taek2222 replied:
답변 감사합니다. 테스트 코드의 양보다는 질적인 면, 꼼꼼한 테스트 케이스에 더 집중해 보겠습니다. 🫡

taek2222 reviewed (commented):
@jinyoungchoi95 
시간 내어 리뷰해 주셔서 정말 감사합니다! ☺️

말씀해 주신 것처럼, 레벨 1에서는 `자바` 학습에 집중하는 것이 중요하다고 생각하여 아키텍처 분리를 최소화하고, 객체만으로 구현할 수 있는 형태로 변경했습니다.

이 과정에서 많은 것을 배웠고, 앞으로 구현 시 주의해야 할 부분도 여러 가지 캐치할 수 있었습니다. 정말 감사합니다. 🙇🏻

또한, 최소한의 TDD 과정을 수행하면서 객체 간 의존성이 높으면 오히려 테스트가 어려워진다는 점을 뼈저리게 느꼈습니다. 앞으로는 이 부분을 더욱 신경 쓰며 개선해 나가겠습니다. ⚙️

정성스러운 피드백 다시 한번 감사드립니다! 부족한 부분이 많지만, 계속해서 고쳐 나가며 성장하고 싶습니다. 

계속해서 부족한 점이 있다면 짚어주시면 감사하겠습니다. 배움을 통해 더욱 발전해 나가겠습니다. 🔥😊

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceStateType.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.util.Arrays;
+
+public enum AttendanceStateType {
```
`AttendanceState`만으로도 의미전달은 충분한 것 아닐까요? enum이 이미 이 클래스의 의미를 전달해주고 있는데 Type이라는 의미가 enum과 중복되지는 않는지 고민해보시면 좋을 것 같아요.

taek2222 replied:
확실히 중복되는 느낌이 있습니다. 👍

이 부분도 인지하고 있었지만, 클래스 명으로도 enum과 같은 의미를 두고자 정의했던 거 같습니다. 개선해 보겠습니다!

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceStateType.java:
```diff
@@ -0,0 +1,28 @@
+public enum AttendanceStateType {
+    ABSENCE("결석", 30),
+    LATE("지각", 5),
+    ATTENDANCE("출석", 0);
```
출력의 요구사항이 영어로 표현되도록 원하면 화면의 요구사항으로 인해 domain 코드의 영향이 있겠네요.

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceStateTypeTest.java:
```diff
@@ -0,0 +1,32 @@
+        return Stream.of(
+                Arguments.of(0, ATTENDANCE),
+                Arguments.of(6, LATE),
+                Arguments.of(31, ABSENCE)
```
임계값에 해당하는 출결 상태를 원한다면 5, 30도 테스트되어야할 것 같아요.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceStateType.java:
```diff
@@ -0,0 +1,28 @@
+        this.threshold = threshold;
+    }
+
+    public static AttendanceStateType find(final int overTime) {
```
-1이 들어오면 어떻게 되나요?

taek2222 replied:
> -1이 들어오면 어떻게 되나요?

A. `출석`으로 반환하게 됩니다. `OverTime`의 값이 추출될 때 `0` 미만의 값은 `0` 으로 반환되게 설정했습니다. 다만, 확장성을 고려했을 때, 해당 메서드가 다른 곳에서도 사용된다면 `OverTime` 계산을 인지하지 못하면 문제가 될 거 라는 생각을 했습니다. 한 번 개선해 보겠습니다.

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceInit.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.List;
+import java.util.stream.IntStream;
+
+public class AttendanceInit {
```
domain인 이유가 있을까요?

taek2222 replied:
여러 `Domain`이 필드로 연결되어 있고, 로직에서 수행하는 부분이 Domain의 영향에 가깝다고 생각해 `Domain` 이라 생각했습니다! 🤔

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,88 @@
+        return List.of(oldAttendance, newAttendance);
+    }
+
+    public List<Attendance> getAttendanceRecord(final LocalDate today, final String nickname) {
```
`Attendances`라는 일급컬랙션이 있는데 `List<Attendance>`를 반한하여 사용하는 메서드들이 있네요. 비타가 정의하는 일급컬랙션은 무엇인가요? 왜 사용하여야하나요?

taek2222 replied:
지적해 주셔서 진심으로 감사합니다. 🙇🏻🙇🏻

일급 컬렉션의 의미를 대충(?) 정말 모호하게 알고 있었던 점을 파악했습니다.
질문을 들었을 때, "일급 컬렉션? 왜 대답을 못하지?" 라는 모습과 함꼐 학습을 시도했습니다. 

물론, 지금도 부족한 경험이 있지만 앞으로 일급컬렉션 적용을 적극적으로 해보면서 가독성 및 메서드 분리 면에서 엄청난 효율을 발휘할 수 있을거라 확신하고 있습니다! 🔥

계속해서 연습하고 적용해 보겠습니다. 현 프로젝트에서는`AttendanceManager` 의 반환되는 것만 일급 컬렉션을 적용했습니다! ☺️

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceManagerTest.java:
```diff
@@ -0,0 +1,315 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+
+class AttendanceManagerTest {
```
가독성이 고민이시라면 `@Nested`를 통해서 동일한 기능들을 테스트하는 메서드들을 묶어볼 수 있겠네요.

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceManagerTest.java:
```diff
@@ -0,0 +1,315 @@
+
+class AttendanceManagerTest {
+
+    private DateGenerator dateGenerator;
+    private AttendanceManager attendanceManager;
+
+    @BeforeEach
+    void beforeEach() {
+        dateGenerator = new MockingDateGenerator();
+        attendanceManager = new AttendanceManager();
+    }
```
```suggestion
    private final DateGenerator dateGenerator = new MockingDateGenerator();
    private final AttendanceManager attendanceManager = new AttendanceManager();
```
junit5의 기본 라이프사이클은 메소드당이기 때문에 굳이 beforeEach로 가변적으로 선언할 필요는 없어요.

https://junit.org/junit5/docs/current/user-guide/#writing-tests-test-instance-lifecycle

taek2222 replied:
헉 감사합니다! 몰랐던,, 사실입니다 😲

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceManagerTest.java:
```diff
@@ -0,0 +1,315 @@
+        );
+    }
+
+    static class MockingDateGenerator implements DateGenerator {
```
Mock이라는 네이밍이 적절한가요? test double 중 다른 적절한 네이밍은 없을까요?

taek2222 replied:
Fake, Stub, Fixed, Test 중에서 의미가 비슷한 것은 `Fixed` 라고 생각했습니다. 특정 날짜로 고정된 형태로 생성하기 때문입니다! 🤔

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceManagerTest.java:
```diff
@@ -0,0 +1,315 @@
+
+        // then
+        assertThat(result.getStatus().size()).isEqualTo(3);
+        assertThat(result.getRiskType()).isEqualTo(AttendanceRiskType.NONE);
```
149번 assertion이 실패해도 150번 assertion에 대한 테스트피드백도 받을 수 있는 방법은 없을까요?

taek2222 replied:
``` java
        assertAll(
                () -> assertThat(result.getStatus().size()).isEqualTo(3),
                () -> assertThat(result.getRisk()).isEqualTo(AttendanceRisk.NONE)
        );
```

assertAll 에 대해서 처음 알게 되었습니다. 감사합니다 ☺️

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceManagerTest.java:
```diff
@@ -0,0 +1,315 @@
+            dateTimes.add(LocalDateTime.of(date, LocalTime.MAX));
+        }
+
+        for (String nickname : nicknames) {
+            Attendances attendances = new Attendances();
+            for (LocalDateTime dateTime : dateTimes) {
+                attendances.addAttendance(dateTime);
+            }
+            attendanceManager.addCrew(nickname, attendances);
+        }
```
`addCrew`로 데이터를 넣어야해서 given 데이터를 받기 위한 과정들이 힘들어보이는데 단순하게 `AttendanceManager`의 생성자에서 받아서 한번에 생성하면 되지 않을까요?

taek2222 replied:
네..! 맞습니다. 

조언해주신 거 처럼 

1. 생성자에서 한 번에 받아서 처리하는 방식 
2. 매개변수로 한 번에 생성할 수 있는 방식

두 가지의 방법 중 하나만 사용했다면 테스트도 용이했을 거 같습니다. 😢

다만, `AttendanceManager`를 `AppConfig` 내에서 생성해 주입하는 방식으로 되어 있어서 전부 수정하기에는 방향성이 잡히지 않습니다.. 😢

`Step 2` 에서는 충분한 고려로 간단하게 작성해보겠습니다!

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private EnumMap<AttendanceStateType, Integer> calculateAttendanceStatus(final List<Attendance> attendances) {
+        EnumMap<AttendanceStateType, Integer> status = new EnumMap<>(AttendanceStateType.class);
+        Arrays.stream(values())
+                .forEach(stateType -> {
+                    int stateCount = calculateStateCount(attendances, stateType);
+                    status.put(stateType, stateCount);
+                });
+        return status;
```
```suggestion
        return Arrays.stream(AttendanceStateType.values())
                .collect(Collectors.toMap(
                        stateType -> stateType,
                        stateType -> calculateStateCount(attendances, stateType),
                        (a, b) -> b,
                        () -> new EnumMap<>(AttendanceStateType.class)
                ));
```
로 표현될 수 있겠네요. 

stream의 foreach 종단 연산은 안티패턴입니다.

https://tecoble.techcourse.co.kr/post/2020-05-14-foreach-vs-forloop/

taek2222 replied:
전혀 모르고 있었던,, 사실입니다. 리뷰 감사합니다! ☺️

jinyoungchoi95 commented on src/main/java/attendance/domain/CampusTime.java:
```diff
@@ -0,0 +1,20 @@
+
+    public static void validateOperationTime(final LocalTime time) {
+        if (time.isBefore(START_TIME.time) || time.isAfter(END_TIME.time)) {
+            throw new IllegalArgumentException("[ERROR] 캠퍼스 운영시간이 아닙니다.");
```
몇시부터 몇시까지가 캠퍼스 운영시간인지 메시지에서 명확하게 전달해주는건 어떨까요?

taek2222 replied:
에러 메시지를 더 명확하게 해보겠습니다! 🫡

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceInit.java:
```diff
@@ -0,0 +1,73 @@
+        this.holiday = holiday;
+    }
+
+    public void initSystem() {
+        initAttendances();
+        initHoliday();
+    }
```
생성자의 역할에서 init을 분리하는 이유가 있을까요? `addHoliday`도 없이 생성자에서만 구성하게 만들 수 있는데 init이 분리되어 오히려 불안정하게 만드는 원인으로 보이네요. 생성한 후 init을 외부에 호출하는데 의존하게되어 init이 호출되지 않으면 어떻게 되나요?

taek2222 replied:
감사합니다! 말씀해주신 것처럼 _**"init을 외부에 호출하는데 의존하게되어 init이 호출되지 않으면 어떻게 되나요?"**_ 와 같은 문제가 생기지 않도록, 생성자에서 처리되도록 개선하겠습니다. 😊

또한, `addHoliday`를 생성한 이유는 초기 페어 프로그래밍 시 실제 프로젝트처럼 확장성을 고려해보자는 의견에서 비롯되었습니다. 긴급한 폭설이나 폭우로 인해 공휴일이 추가될 수 있는 상황을 가정하고 구현한 기능입니다!

jinyoungchoi95 commented on src/main/java/attendance/utility/CurrentDateGeneratorImpl.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.time.LocalDate;
+
+public class CurrentDateGeneratorImpl implements DateGenerator {
+
+    @Override
+    public LocalDate now() {
+        return LocalDate.now();
+    }
```
`DateGenerator`로 추상화했는데 메소드명이 명시적인 `now()`이네요. 충분히 추상화가 된 것 맞나요?

jinyoungchoi95 commented on src/main/java/attendance/AttendanceMenu.java:
화면에 대한 정보이니 `view`에 위치하는 것이 더 적절해보이네요.

taek2222 replied:
확실히 `View` 레이어에 가까워 보입니다! 새로운 시선을 볼 수 있었습니다! ☺️

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,88 @@
+        return crewAttendances.get(nickname);
+    }
+
+    private LinkedHashMap<String, AttendanceStatus> generateAttendanceRisks(final LocalDate today) {
```
```suggestion
    private Map<String, AttendanceStatus> generateAttendanceRisks(final LocalDate today) {
```

taek2222 replied:
감사합니다..! 🙈

jinyoungchoi95 commented on src/main/java/attendance/view/InputView.java:
```diff
@@ -0,0 +1,45 @@
+        return console.nextLine();
+    }
+
+    public String readNickname(boolean isForUpdated) {
+        if (!isForUpdated) {
+            System.out.println(NEW_LINE + "닉네임을 입력해 주세요.");
+        }
+
+        if (isForUpdated) {
+            System.out.println(NEW_LINE + "출석을 수정하려는 크루의 닉네임을 입력해 주세요.");
+        }
+        return console.nextLine();
```
`console.nextLine()`을 같이 사용한다고 해도 두 기능은 전혀 다른 기능입니다. boolean으로 분기하지 말고 사용처에 따라 명확하게 메소드를 분리해서 사용하세요.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 비타 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

taek2222 committed:
refactor(InputView): 일반, 업데이트 입력 분리

taek2222 committed:
refactor(AttendanceManager): 반환 타입 Map으로 변경

taek2222 committed:
refactor(AttendanceMenu): 레이어 변경

taek2222 committed:
refactor(CampusTime): 캠퍼스 에러 메시지 더욱 명확하게 변경

taek2222 committed:
feat(AttendanceSystem): 시간 검증하는 메서드 기능 추가

taek2222 committed:
fix(AttendanceStatus): 종단 안티 패턴 개선

taek2222 committed:
refactor(AttendanceRisk, AttendanceState): 네이밍의 Type 제외

taek2222 committed:
test(AttendanceStateTest): 테스트 파라미터 추가

- 임계값에 해당되는 부분을 추가적으로 테스트하는 로직을 추가했습니다.

taek2222 committed:
refactor(DateGenerator): 추상화 메서드 네이밍 변경

taek2222 committed:
test(AttendanceManagerTest): 테스트 리팩토링

- Mock 네이밍 보다는 고정된 느낌이 강해 Fixed로 변경했습니다.

taek2222 committed:
test(*): 테스트 리팩토링

- 여러 검증을 한 번에 assertAll 로 묶어서 관리하도록 개선했습니다.

taek2222 committed:
refactor(AttendanceState): 오버 시간 계산 로직 병합

taek2222 committed:
refactor(AttendanceInit): 생성자로 바로 초기화 할 수 있도록 변경

taek2222 committed:
refactor(AppConfig): 사용하지 않는 필드 정리

taek2222 committed:
refactor(AttendanceUpdate): 업데이트 일급 컬렉션 정의

taek2222 committed:
refactor(Attendances): 정적 팩토리 메서드 적용 및 일급 컬렉션 반환 변경

taek2222 committed:
refactor(AttendanceRiskCrews): 출석 위험자 크루 일급컬렉션 적용

taek2222 committed:
refactor(AttendanceSystem): 메서드 네이밍 간소화

taek2222 committed:
feat(AttendanceStateView): view 레이어의 출석 출력 추가

taek2222 committed:
refactor(AttendanceState): 필드의 출력 관련 제거

taek2222 committed:
refactor(AttendanceStateView): 에러 메시지 개선

taek2222 committed:
feat(AttendanceRiskView): 위험도 View 추가

taek2222 committed:
refactor(AttendanceRisk): 위험도 View 분리로 이름 삭제

taek2222 committed:
refactor(OutputView): Enum View 분리로 로직 개선

taek2222 reviewed (commented):
두 번째 리뷰 요청도 꼼꼼하게 봐주셔서 감사합니다. 덕분에 또 여러 가지를 배울 수 있었습니다! 📚 (아직 부족하지만.. 😂)

이번에 두 번째 코멘트를 적용하면서 몇 가지 궁금한 점이 생겨 질문을 남깁니다.

> Q. 리팩토링을 하다 보면 코드의 변경점이 생기고, 이에 따라 테스트 코드도 수정이 필요한데, 이 경우 커밋 단위를 나누는 게 좋을까요?
(예: test: ~~, refactor: ~~ 형태로 구분하는 방식)

> Q. TDD로 첫 로직을 구성할 때, 임시 테스트 클래스를 먼저 생성해서 진행하나요? 아니면 대략적인 설계를 먼저 한 뒤, 클래스를 만들고 해당 클래스의 테스트 클래스를 작성하는 방식이 일반적인가요?

jinyoungchoi95 commented on src/test/java/attendance/domain/HolidayTest.java:
> Q. 리팩토링을 하다 보면 코드의 변경점이 생기고, 이에 따라 테스트 코드도 수정이 필요한데, 이 경우 커밋 단위를 나누는 게 좋을까요?
(예: test: ~~, refactor: ~~ 형태로 구분하는 방식)

이름 변경으로 테스트 코드가 변경되는 것이 아니고 리팩터링을 통해서 테스트코드가 수정이 필요하면 그건 기능변경이지 리팩터링이 아닙니다.

jinyoungchoi95 commented on src/test/java/attendance/domain/CampusTimeTest.java:
> TDD로 첫 로직을 구성할 때, 임시 테스트 클래스를 먼저 생성해서 진행하나요? 아니면 대략적인 설계를 먼저 한 뒤, 클래스를 만들고 해당 클래스의 테스트 클래스를 작성하는 방식이 일반적인가요?

개발을 함에 있어서 일반적이다라는 것은 없습니다. 누구는 전자를 하고 누구는 후자의 방법을 나름대로의 이유에 따라서하는거죠. 우테코 내에서는 일반적인 방법, 정답 등을 찾지 않으셨으면 좋겠어요.

TDD로 첫 로직을 구성하는데 어려움이 있다는 것은 무엇을 해야할지 정하지않고 무작정 시작하려하기 때문이에요. 개발을 시작하기전에 개발과 관련된 모든 것을 치우고 종이하나로 설계를 하루종일 해보고 작업해보면 하던 방법과의 차이를 느끼실 수 있을 것 같네요.

taek2222 replied:
좋은 말씀 감사합니다 🙇🏻 _**"일반적인 방법, 정답 등을 찾지 않으셨으면 좋겠어요."**_ 가장 와닿고 기억해두겠습니다!

jinyoungchoi95 commented on src/main/java/attendance/domain/AttendanceInit.java:
해당 클래스에 대해 코멘트를 드렸던 이유는 초기화를 담당하는 클래스가 존재 의미가 없기 때문입니다. 생성로직을 한번 더 래핑한건데 그게 의미가 있는 것처럼 보이진 않습니다.

객체가 객체답게 본인의 역할이 있도록 설계하세요. 해당 클래스는 본인이 다른 객체들과 유기적으로 결합하여 상호작용을 하고 있나요?

jinyoungchoi95 commented on src/main/java/attendance/utility/DateGenerator.java:
```diff
@@ -0,0 +1,7 @@
+import java.time.LocalDate;
+
+public interface DateGenerator {
+    LocalDate generateNow();
```
now라는 이름이 문제였기때문에 말씀드린 코멘트였습니다. 이 메소드를 구현한 구현체는 무얼하든 현재를 제공하게 되는 것 아닐까요?

jinyoungchoi95 commented on src/main/java/attendance/AttendanceSystem.java:
```diff
@@ -0,0 +1,141 @@
+import static attendance.view.AttendanceMenu.RISK_SEARCH;
+import static attendance.view.AttendanceMenu.UPDATE;
+
+public class AttendanceSystem {
```
`AttendanceSystem`과 `AppConfig`가 분리될 필요가 없어보이네요.

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceManagerTest.java:
```diff
@@ -0,0 +1,314 @@
+        attendanceManager.addCrew(nickname, attendances);
+
+        // when
+        LocalDateTime dateTime = LocalDateTime.of(nowDate, time);
```
when에는 수행하고자하는 행위가 들어가야하므로 dateTime은 제공되는 값인 given에 해당합니다.

jinyoungchoi95 commented on src/test/java/attendance/domain/AttendanceStatusTest.java:
```diff
@@ -0,0 +1,167 @@
+        ));
+
+        // when
+        AttendanceStatus originStatus = new AttendanceStatus(origin);
+        AttendanceStatus comparedStatus = new AttendanceStatus(compared);
+
+        // then
+        assertThat(originStatus.compareTo(comparedStatus))
+                .isPositive();
```
마찬가지로 테스트 대상은 `compareTo`이므로 when에 있는 값들은 모두 given에 해당하며 `compareTo`가 when에 해당합니다.

jinyoungchoi95 reviewed (approved):
안녕하세요 비타 😄

현재 단계에 필요한 내용들은 충분히 확인한 것 같아 다음 미션 진행부탁드려요.

그리고 이후부터는 코멘트의 resolve 처리는 하지 않으셔도 됩니다. 반영 여부를 리뷰어가 확인하기 힘들뿐더러 pr 코멘트들은 다른 사람 혹은 나중에 비타도 보게되는 학습자료가 되므로 미션내에서는 resolve 처리는 하지 않는 것이 좋습니다.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다. 
감사합니다 🙇‍♂️
